### PR TITLE
docs: add automated upgrades proposal

### PR DESCRIPTION
**Pull request by andrewrynhard**
This is a proposal for the new way of running Talos and performing upgrades in v0.2

Signed-off-by: Andrew Rynhard <andrew@andrewrynhard.com>

### CODE REVIEW

File: docs/proposals/20190704-RunModeAndUpgrades.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+Our current design, along with our immutable approach, necessitates two root partitions in order to perform an upgrade.
+These two partitions are merely toggled on each upgrade (e.g. When upgrading while running `ROOT-A`, the `ROOT-B` partition if used for the new install and then targeted on the next boot).
+This design makes our footprint bigger than we need.
+
+## Proposal
+
+- use squashfs
+- run purely in initramfs
+- reduce rootfs.tar.gz down to cached containers
+- use container registry technology to perform installs and upgrades
```
Comment (bradbeam) [inline docs/proposals/20190704-RunModeAndUpgrades.md]: I get what you're getting at with this, but I don't it's conveying what you're wanting to get at. 
Perhaps something like

> Distribute upgrade code and assets as a container

File: docs/proposals/20190704-AutomatedUpgrades.md [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+### Automated Upgrades
+
+All of the aformentioned improvements help in the higher goal of automated upgrades.
+Once the above is implemented, upgrades should be simple and self-contained to a machine.
+By running upgrades in containers, we can have very simple operator implementation that can rollout an upgrade across a cluster by simply:
+
+- checking for a new tag
+- calling the appropriate API
+- orchestrating cluster-wide rollout
+
```
Comment (andrewrynhard) [inline docs/proposals/20190704-AutomatedUpgrades.md]: This needs to mention the Kubernetes operator.

File: docs/proposals/20190704-AutomatedUpgrades.md [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+We currently do not compress the container tarballs.
+The idea here is that by compressing the container images, and then creating a `squashfs`, we can reduce the `rootfs` down to an acceptable size.
+This has the advantage that everything is nicely bundled together.
+
+A downside is that it will take the kernel longer to decompress the `initramfs`.
+Another downside is that we would require at least 1GB of RAM to run talos.
+
+#### Proposal 3
+
+Finally, we could also remove the Kubernetes tarballs entirely and let the `kubelet` download them.
```
Comment (smira) [inline docs/proposals/20190704-AutomatedUpgrades.md]: can we download approved tarball from some Talos location? like customer might override the URL, but we verify the checksum?

File: docs/proposals/20190704-AutomatedUpgrades.md [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+With the move to run the `rootfs` in memory, we will remove the need for `ROOT-A` and `ROOT-B` partitions.
+In the case of cloud installs, or where users may want to boot without PXE/iPXE, we will require a boot partition.
+In all cases we require a data partition.
+
+I also propose we rename the `DATA` partition to `EPHEMERAL` to make it clear that this partition should not be used to store application data.
+Our suggestion to users who do want persistent storage on the host is to mount another disk and to use that.
+Talos will make the guarantee to only manage the "owned" block device.
+
+### Performing Upgrades in a Container
```
Comment (andrewrynhard) [inline docs/proposals/20190704-AutomatedUpgrades.md]: An alternative we discussed in a community meeting is to _not_ run upgrades in a container, but to still used a container image to deliver the artifacts. This will remove the need for `containerd` in the early boot environment.

File: docs/proposals/20190704-AutomatedUpgrades.md [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+For performance reasons, and a smaller footprint, we can create our `rootfs` as a `squashfs` image.
+The way in which we can run the `rootfs` is by mounting it using a loopback device.
+Once mounted, we have will perform a `switch_root` into the mounted `rootfs`.
+
+> Note: A benefit of `squashfs` is that it allows us to retain our read only `rootfs` without having to mount it as such.
+
+### Slimming down `rootfs`
+
+To slim down the size of our rootfs, I propose that we remove Kubernetes tarballs entirely and let the `kubelet` download them.
+This is a shift from our immutability goals, but in this case we can embrace configurability at the Kubernetes layer.
```
Comment (Ulexus) [inline docs/proposals/20190704-AutomatedUpgrades.md]: This doesn't really shift immutability.  squashfs is still intrinsically immutable.  We're just decoupling the kubelet image from the rootfs image.  While that does _allow_ a flexion point, it needn't _mandate_ that flexion point.

File: docs/proposals/20190704-AutomatedUpgrades.md [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+### Automated Upgrades
+
+All of the aformentioned improvements help in the higher goal of automated upgrades.
+Once the above is implemented, upgrades should be simple and self-contained to a machine.
+By running upgrades in containers, we can have very simple operator implementation that can rollout an upgrade across a cluster by simply:
+
+- checking for a new tag
+- orchestrating cluster-wide rollout
+
+The operator will initially run upgrades only, but it will also provide a basis for a number of future automation opportunities.
```
Comment (Ulexus) [inline docs/proposals/20190704-AutomatedUpgrades.md]: Perhaps "coordinate upgrades" rather than "run upgrades" would be more accurate?

File: docs/proposals/20190704-AutomatedUpgrades.md (hunk 5 above)
Comment (andrewrynhard) [reply docs/proposals/20190704-AutomatedUpgrades.md]: Agree. I can reword.

File: docs/proposals/20190704-AutomatedUpgrades.md (hunk 6 above)
Comment (andrewrynhard) [reply docs/proposals/20190704-AutomatedUpgrades.md]: This is a good point. I don’t we are entirely sure or we haven’t discussed at depth how upgrades would work from the operator point of view. I guess there are two ways. One is the way you are suggesting. If I understand this suggestion it would work by having each individual node invoke their own upgrade but perhaps getting a lock to do so. The alternative is that the operator invoking the node’s API and manages the scheduling internally. Do you see benefits of one over the other? I plan on writing a separate proposal, or maybe even expanding this one, about how this would work so good if we start the discussion now.

Comment (Ulexus) [reply docs/proposals/20190704-AutomatedUpgrades.md]: Perhaps I am being too pedantic, but both of those are examples of the operator "coordinating" the upgrades, by my meaning.  I don't think the former is a good idea, topologically.  That is, I don't think the talos components should reach into kubernetes to acquire the lock.  
The flow should, in my opinion flow from the operator to talos, only.  Thus, the scheduling, locking, etc, would all be under the strict purview of the operator.

Basically, in the ideal case, an operator should act as a human, using the same tools and APIs as a human, just in an automated fashion.

Comment (Ulexus) [reply docs/proposals/20190704-AutomatedUpgrades.md]: It's the problem with text reviews.  I start focusing on the English rather than the engineering intention.

Comment (andrewrynhard) [reply docs/proposals/20190704-AutomatedUpgrades.md]: > Basically, in the ideal case, an operator should act as a human, using the same tools and APIs as a human, just in an automated fashion.

Agree with this 100%. I will add some more to this proposal around how the operator will work. 

> It's the problem with text reviews. I start focusing on the English rather than the engineering intention.

My bad English probably doesn’t help.

Comment (andrewrynhard) [reply docs/proposals/20190704-AutomatedUpgrades.md]: Added a section that hopefully capture this.

File: docs/proposals/20190704-AutomatedUpgrades.md [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+- beta
+- stable
+
+The implementation details around this service and its' interaction with the operator is not in the scope of this proposal.
+
+#### Workflow
+
+In order to upgrade nodes, the operator will subscribe to events for nodes.
+It will keep an in-memory list of all nodes.
+The operator will periodically query the target registry for a list of containers within a time range.
```
Comment (andrewrynhard) [inline docs/proposals/20190704-AutomatedUpgrades.md]: Separate paragraph.

File: docs/proposals/20190704-AutomatedUpgrades.md [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+The time range should start where a previous query has ended.
```
Comment (andrewrynhard) [inline docs/proposals/20190704-AutomatedUpgrades.md]: The initial query should use the currently installed image's timestamp to decide where to start. This means we should add this as a label.

File: docs/proposals/20190704-AutomatedUpgrades.md [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+### Automated Upgrades
+
+All of the aformentioned improvements help in the higher goal of automated upgrades.
+Once the above is implemented, upgrades should be simple and self-contained to a machine.
+By running upgrades in containers, we can have very simple operator implementation that can rollout an upgrade across a cluster by simply:
+
+- checking for a new tag
+- orchestrating cluster-wide rollout
+
+The operator will initially coordinate upgrades only, but it will also provide a basis for a number of future automation opportunities.
```
Comment (Ulexus) [inline docs/proposals/20190704-AutomatedUpgrades.md]: One implementation to note for the operator:  unlike the lower-level upgrade executor, the operator has available to it the drain/cordon action at the k8s level.

Comment (andrewrynhard) [reply docs/proposals/20190704-AutomatedUpgrades.md]: Come to find out, all kubelets have the permissions to cordon and drain themselves, and that is actually part of the upgrade RPC today. Interesting that CLUO opted not to do this 🤔. I can't think of a good reason _not_ to.

Comment (Ulexus) [reply docs/proposals/20190704-AutomatedUpgrades.md]: I can't say for certain that they did not.  I didn't look at the code.  It just doesn't appear to operate as if it did.

File: docs/proposals/20190704-AutomatedUpgrades.md [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+1. an official release
+2. within a given channel
+3. contains a version that is compatible with the current cluster
+
+There may be more information we end up needing, the above is just an example.
+
+Once a container has been chosen, the operator will then create a schedule based on polices.
+The initial policies will perform an upgrade based on:
+
+1. x% of nodes at a given time
```
Comment (Ulexus) [inline docs/proposals/20190704-AutomatedUpgrades.md]: I'd recommend using discrete node counts rather than percentages.  In general, those are better understood and more in-line with dimensioning for all but the largest of clusters.

File: docs/proposals/20190704-AutomatedUpgrades.md [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+    - Replace
+
+A user should be able to apply more than one policy.
+
+Once the operator has decided scheduling, it will use the upgrade RPC to perform an upgrades according to the calculated upgrade schedule.
+
+## Rationale
+
+One of the goals in Talos is to allow for machines to be thought of as a container in that it is ephemeral, reproducable, and predictable.
+Since the `kubelet` can reproduce a machine's workload, it is not too crazy to think that we can facilitate upgrades by completely wiping a node, and then performing a fresh install as if the machine is new.
```
Comment (Ulexus) [inline docs/proposals/20190704-AutomatedUpgrades.md]: ...with special handling for master (etcd) nodes

File: docs/proposals/20190704-AutomatedUpgrades.md (hunk 10 above)
Comment (andrewrynhard) [reply docs/proposals/20190704-AutomatedUpgrades.md]: Good point. I think I might have this covered a few bullet points down. There is an extra time element in my proposal but we could easily have a case where users can specify a time of 0, indicating that a rollout should happen one after the other.

File: docs/proposals/20190704-AutomatedUpgrades.md (hunk 11 above)
Comment (andrewrynhard) [reply docs/proposals/20190704-AutomatedUpgrades.md]: Glad you said something. Today we perform upgrades of masters by wiping the Etcd data, with the idea that etcd _should_ be able to rebuild itself. The plan was to keep this. and add extra checks when upgrading masters that would ensure etcd is caught up before moving on and upgrading the next master. This could qualify as the "special handling" but curious to get your thoughts on this. The alternative is that we perform a snapshot and restore the snapshot.

Comment (Ulexus) [reply docs/proposals/20190704-AutomatedUpgrades.md]: I'm not fundamentally opposed to the idea, but it is definitely something to be _sure_ to get right.  Regardless of other cluster-wide upgrade settings for node-down tolerance, for example, only one master should be down at any given time and _ideally_ no worker node at the same time (to reduce the change load on the etcd cluster).

I think snapshotting and restoring from snapshot is actually more fraught.  If any of the identifying information for the master node is changed, you can end up with some... not split brain, exactly, but non-agreeing information about the way to contact that peer.  It's better (IMO) to remove the node from membership and then re-add it.

In an ideal world, we would bring up a worker node as a non-voting (learner) etcd member, wait for it to be synchronized, remove-add membership, wait for the master to upgrade, set it to non-voting/learning, then remove-add membership again.

This is a feature of etcd 3.4 specifically to streamline membership changes... however, involving a worker node might result in any number of potential complications, so I don't know that that will ultimately be a _good idea.

Minimally, I think we should make a snapshot before the each node upgrade just in case, but not use that backup unless catastrophe strikes.
